File: src/main/java/com/kgaft/VulkanContext/Vulkan/VulkanInstance.java
package com.kgaft.VulkanContext.Vulkan;

import com.kgaft.VulkanContext.Exceptions.BuilderNotPopulatedException;
import com.kgaft.VulkanContext.Exceptions.NotSupportedExtensionException;
import com.kgaft.VulkanContext.Exceptions.NotSupportedLayerException;
import com.kgaft.VulkanContext.MemoryUtils.DestroyableObject;
import com.kgaft.VulkanContext.Vulkan.VulkanLogger.VulkanLogger;
import org.lwjgl.PointerBuffer;
import org.lwjgl.system.MemoryStack;
import org.lwjgl.vulkan.*;

import java.nio.ByteBuffer;
import java.util.List;

import static org.lwjgl.vulkan.VK13.*;

public class VulkanInstance extends DestroyableObject {
    public static final String VK_LAYER_KHRONOS_validation = "VK_LAYER_KHRONOS_validation";
    public static final String VK_LAYER_KHRONOS_synchronization2 = "VK_LAYER_KHRONOS_synchronization2";
    public static final String VK_LAYER_KHRONOS_profiles = "VK_LAYER_KHRONOS_profiles";
    public static final String VK_LAYER_LUNARG_api_dump = "VK_LAYER_LUNARG_api_dump";
    public static final String VK_LAYER_LUNARG_gfxreconstruct = "VK_LAYER_LUNARG_gfxreconstruct";
    public static final String VK_LAYER_LUNARG_monitor = "VK_LAYER_LUNARG_monitor";
    public static final String VK_LAYER_LUNARG_screenshot = "VK_LAYER_LUNARG_screenshot";
    private static VulkanInstanceBuilder builderInstance = null;

    public static VulkanInstanceBuilder getBuilderInstance() {
        if (builderInstance == null) {
            builderInstance = new VulkanInstanceBuilder();
        }
        return builderInstance;
    }

    public static VulkanInstance createInstance(int[] resOutput)
            throws BuilderNotPopulatedException, NotSupportedExtensionException, NotSupportedLayerException {
        if (!builderInstance.appInfoEnabled) {
            throw new BuilderNotPopulatedException("Error you have not specified the info about your application");
        }
        try (MemoryStack stack = MemoryStack.stackPush()) {
            try {
                if (!builderInstance.enabledExtensions.isEmpty()) {
                    checkExtensions(builderInstance.enabledExtensions, stack);
                    builderInstance.createInfo
                            .ppEnabledExtensionNames(stack.callocPointer(builderInstance.enabledExtensions.size()));
                    builderInstance.enabledExtensions.forEach(element -> {
                        builderInstance.createInfo.ppEnabledExtensionNames().put(stack.UTF8Safe(element));
                    });
                    builderInstance.createInfo.ppEnabledExtensionNames().rewind();
                }
                if (!builderInstance.enabledLayers.isEmpty()) {
                    checkLayers(builderInstance.enabledLayers, stack);
                    builderInstance.createInfo
                            .ppEnabledLayerNames(stack.callocPointer(builderInstance.enabledLayers.size()));
                    builderInstance.enabledLayers.forEach(element -> {
                        builderInstance.createInfo.ppEnabledLayerNames().put(stack.UTF8Safe(element));
                    });
                    builderInstance.createInfo.ppEnabledLayerNames().rewind();
                }
            } catch (NotSupportedExtensionException | NotSupportedLayerException e) {
                
                throw e;
            }
            VulkanLogger logger = new VulkanLogger();
            VkDebugUtilsMessengerCreateInfoEXT debugCreateInfo = null;
            if (!builderInstance.enabledLayers.isEmpty()) {
                debugCreateInfo = VkDebugUtilsMessengerCreateInfoEXT.calloc(stack);
                logger.describeLogger(stack, debugCreateInfo);
                builderInstance.createInfo.pNext(debugCreateInfo);
            }
            PointerBuffer instRes = stack.callocPointer(1);
            int res = vkCreateInstance(builderInstance.createInfo, null, instRes);
            if (resOutput != null) {
                resOutput[0] = res;
            }
            if (res != VK_SUCCESS) {
                return null;
            }
            VkInstance instance = new VkInstance(instRes.get(), builderInstance.createInfo);
            VulkanInstance result = new VulkanInstance(instance, null);
            if (!builderInstance.enabledLayers.isEmpty()) {
                logger.initLoggerInstance(instance, debugCreateInfo, stack);
                result.logger = logger;
            }
            result.setEnabledLayers(builderInstance.enabledLayers);

            return result;
        }

    }

    private static void checkLayers(List<String> toCheck, MemoryStack stack) throws NotSupportedLayerException {
        int[] layerCount = new int[1];
        vkEnumerateInstanceLayerProperties(layerCount, null);
        VkLayerProperties.Buffer layers = VkLayerProperties.calloc(layerCount[0], stack);
        vkEnumerateInstanceLayerProperties(layerCount, layers);
        for (String el : toCheck) {
            boolean found = false;
            for (VkLayerProperties cel : layers) {
                if (cel.layerNameString().equals(el)) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                throw new NotSupportedLayerException(el);
            }
        }
    }

    private static void checkExtensions(List<String> toCheck, MemoryStack stack) throws NotSupportedExtensionException {
        int[] extensionCount = new int[1];
        vkEnumerateInstanceExtensionProperties((ByteBuffer) null, extensionCount, null);
        VkExtensionProperties.Buffer extensions = VkExtensionProperties.calloc(extensionCount[0], stack);
        vkEnumerateInstanceExtensionProperties((ByteBuffer) null, extensionCount, extensions);
        for (String el : toCheck) {
            boolean found = false;
            for (VkExtensionProperties cel : extensions) {
                if (cel.extensionNameString().equals(el)) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                throw new NotSupportedExtensionException(el);
            }
        }
    }

    private VkInstance instance;
    private VulkanLogger logger;
    private List<String> enabledLayers;

    public VulkanInstance(VkInstance instance, VulkanLogger logger) {
        this.instance = instance;
        this.logger = logger;
    }

    protected void setEnabledLayers(List<String> enabledLayers) {
        this.enabledLayers = enabledLayers;
    }

    public List<String> getEnabledLayers() {
        return enabledLayers;
    }

    public VkInstance getInstance() {
        return instance;
    }

    public VulkanLogger getLogger() {
        return logger;
    }

    @Override
    public void destroy() {
        this.destroyed = true;
        logger.destroy();
        vkDestroyInstance(instance, null);
    }
}

File: src/cxx/VulkanContext/Vulkan/VulkanInstance.cpp
//
// Created by kgaft on 7/26/23.
//


#include "VulkanInstance.h"
#include <cstring>
#include "VulkanContext/Exceptions/BuilderNotPopulatedException.h"
#include "VulkanContext/Exceptions/NotSupportedExtensionException.h"
#include "VulkanContext/Exceptions/NotSupportedLayerException.h"

VulkanInstanceBuilder *VulkanInstance::getBuilderInstance() {
    if (builderInstance == nullptr) {
        builderInstance = new VulkanInstanceBuilder();
    }

    return builderInstance;
}

VkResult VulkanInstance::createInstance(VulkanInstance **pOutput) {
    if (builderInstance->appInfoEnabled) {
        if(!builderInstance->enabledExtensions.empty()){
            checkExtensions(builderInstance->enabledExtensions);
        }
        if(!builderInstance->enabledLayers.empty()){
            checkLayers(builderInstance->enabledLayers);
        }

        VkDebugUtilsMessengerCreateInfoEXT debugCreateInfo = {};
        if (!builderInstance->enabledLayers.empty()) {
            VulkanLogger::describeLogger(debugCreateInfo);
            builderInstance->createInfo.pNext = &debugCreateInfo;
        }
        VkInstance instance;
        VkResult res = vkCreateInstance(&builderInstance->createInfo, nullptr, &instance);
        if(res!=VK_SUCCESS){
            return res;
        }
        if(!builderInstance->enabledLayers.empty()){
            (*pOutput)->logger = new VulkanLogger;
            (*pOutput)->logger->init(instance);
        }
        *pOutput = new VulkanInstance();
        (*pOutput)->instance = instance;
        return res;
    }
    throw BuilderNotPopulatedException("Error, you have not specified application info");
}

void VulkanInstance::checkExtensions(std::vector<const char *> &toCheck) {
    uint32_t extensionCount = 0;
    vkEnumerateInstanceExtensionProperties(nullptr, &extensionCount, nullptr);
    std::vector<VkExtensionProperties> extensions(extensionCount);
    vkEnumerateInstanceExtensionProperties(nullptr, &extensionCount, extensions.data());
    for (auto &el : toCheck) {
        bool found = false;
        for (auto &cel : extensions) {
            if (!strcmp(cel.extensionName, el)) {
                found = true;
                break;
            }
        }
        if (!found) {
            throw NotSupportedExtensionException(el);
        }
    }
}

void VulkanInstance::checkLayers(std::vector<const char *> &toCheck) {
    uint32_t layerCount = 0;
    vkEnumerateInstanceLayerProperties(&layerCount, nullptr);
    std::vector<VkLayerProperties> layers(layerCount);
    vkEnumerateInstanceLayerProperties(&layerCount, layers.data());
    for (auto &el : toCheck) {
        bool found = false;
        for (auto &cel : layers) {
            if (!strcmp(cel.layerName, el)) {
                found = true;
                break;
            }
        }
        if (!found) {
            throw NotSupportedLayerException(el);
        }
    }
}

void VulkanInstance::destroy() {
    delete logger;
    vkDestroyInstance(instance, nullptr);
    destroyed = true;
}


VulkanInstanceBuilder::VulkanInstanceBuilder() {
    createInfo.pApplicationInfo = &appInfo;
    createInfo.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
    createInfo.pNext = nullptr;
}

VulkanInstanceBuilder *
VulkanInstanceBuilder::setApplicationInfo(const char *appName, const char *engineName, uint32_t apiVersion,
                                          uint32_t appVersion, uint32_t engineVersion) {
    appInfoEnabled = true;
    appInfo.sType = VK_STRUCTURE_TYPE_APPLICATION_INFO;
    appInfo.apiVersion = apiVersion;
    appInfo.engineVersion = engineVersion;
    appInfo.applicationVersion = appVersion;
    appInfo.pApplicationName = appName;
    appInfo.pEngineName = engineName;
    appInfo.pNext = nullptr;
    return this;
}

VulkanInstanceBuilder *VulkanInstanceBuilder::addExtension(const char *extension) {
    enabledExtensions.push_back(extension);
    createInfo.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
    createInfo.ppEnabledExtensionNames = enabledExtensions.data();
    createInfo.enabledExtensionCount = enabledExtensions.size();
    createInfo.pNext = nullptr;

    return this;
}

VulkanInstanceBuilder *VulkanInstanceBuilder::addLayer(const char *layer) {
    enabledLayers.push_back(layer);
    createInfo.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
    createInfo.ppEnabledLayerNames = enabledLayers.data();
    createInfo.enabledLayerCount = enabledLayers.size();
    createInfo.pNext = nullptr;
    return this;
}

VulkanInstanceBuilder *VulkanInstanceBuilder::setInstanceFlags(VkInstanceCreateFlags flags) {
    createInfo.flags = flags;
    createInfo.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
    createInfo.pNext = nullptr;
    return this;
}


